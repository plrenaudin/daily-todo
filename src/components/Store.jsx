import React from "react";
import {
  LOAD,
  ADD,
  addTodo,
  REMOVE,
  removeTodo,
  listTodo,
  TOGGLE,
  toggleTodo,
  listWeekTodo
} from "../modules/model";
import { isoFormatDate, currentWeekISO } from "../modules/date";

const ConfigStore = React.createContext();
const TodoStore = React.createContext();
const WeekStore = React.createContext();

const todoReducer = (state, action) => {
  switch (action.type) {
    case LOAD:
      return action.data;
    case ADD:
      return addTodo(state, action.data);
    case REMOVE:
      return removeTodo(state, action.data);
    default:
      throw new Error();
  }
};

const weekReducer = (state, action) => {
  switch (action.type) {
    case LOAD:
      return action.data;
    case TOGGLE:
      return toggleTodo(state, action.todo, action.day);
    default:
      throw new Error();
  }
};

const useDayTodos = () => {
  const [week, dispatch] = React.useReducer(weekReducer, []);
  return { week, dispatch };
};

const useTodos = () => {
  const [todos, dispatch] = React.useReducer(todoReducer, []);
  return { todos, dispatch };
};

const Provider = ({ children }) => {
  //Config
  const [currenDate, setCurrentDate] = React.useState(isoFormatDate());
  const configStore = { currenDate, setCurrentDate };

  //Week
  const weekStore = useDayTodos();

  //Todos
  const todoStore = useTodos();

  React.useEffect(() => {
    const week = currentWeekISO();
    Promise.all([listTodo(), listWeekTodo(week)]).then(([todos, weekTodos]) => {
      todoStore.dispatch({ type: LOAD, data: todos });
      weekStore.dispatch({ type: LOAD, data: weekTodos });
    });
  }, []);

  return (
    <ConfigStore.Provider value={configStore}>
      <TodoStore.Provider value={todoStore}>
        <WeekStore.Provider value={weekStore}>{children}</WeekStore.Provider>
      </TodoStore.Provider>
    </ConfigStore.Provider>
  );
};

export { ConfigStore, WeekStore, TodoStore, Provider };
