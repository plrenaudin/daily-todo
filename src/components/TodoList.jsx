import React from "react";
import { REMOVE } from "../modules/model";
import { TodoStore } from "./Store";

const TodoList = React.memo(() => {
  const { todos, dispatch } = React.useContext(TodoStore);

  const remove = data => dispatch({ type: REMOVE, data });

  return (
    <ul>
      {todos.map(i => (
        <li key={i}>
          {i} <button onClick={() => remove(i)}>X</button>
        </li>
      ))}
    </ul>
  );
});

export { TodoList };
