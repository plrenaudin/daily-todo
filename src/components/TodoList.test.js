import React from "react";
import renderer from "react-test-renderer";
import { TodoList } from "./TodoList";
import "jest-styled-components";
import { TodoStore } from "./Store";
jest.mock("idb");

it("renders the Todolist correctly", () => {
  const component = (
    <TodoStore.Provider value={{ todos: ["plip", "huhu"] }}>
      <TodoList />
    </TodoStore.Provider>
  );
  expect(renderer.create(component).toJSON()).toMatchSnapshot();
});
