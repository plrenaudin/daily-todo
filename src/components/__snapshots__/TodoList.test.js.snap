// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the Todolist correctly 1`] = `
<ul>
  <li>
    plip
     
    <button
      onClick={[Function]}
    >
      X
    </button>
  </li>
  <li>
    huhu
     
    <button
      onClick={[Function]}
    >
      X
    </button>
  </li>
</ul>
`;
